osu.Game.Rulesets.Taiko/Mods/TaikoModFlashlight.cs

```diff
@@ -1,12 +1,77 @@
 ﻿// Copyright (c) 2007-2018 ppy Pty Ltd <contact@ppy.sh>.
 // Licensed under the MIT Licence - https://raw.githubusercontent.com/ppy/osu/master/LICENCE
 
+using osu.Framework.Graphics;
 using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Taiko.Objects;
+using osu.Game.Rulesets.Taiko.UI;
+using osu.Game.Rulesets.UI;
+using OpenTK;
 
 namespace osu.Game.Rulesets.Taiko.Mods
 {
-    public class TaikoModFlashlight : ModFlashlight
+    public class TaikoModFlashlight : ModFlashlight<TaikoHitObject>
     {
         public override double ScoreMultiplier => 1.12;
+
+        private const float default_flashlight_size = 250;
+
+        public override Flashlight CreateFlashlight() => new TaikoFlashlight(playfield);
+
+        private TaikoPlayfield playfield;
+
+        public override void ApplyToRulesetContainer(RulesetContainer<TaikoHitObject> rulesetContainer)
+        {
+            playfield = (TaikoPlayfield)rulesetContainer.Playfield;
+            base.ApplyToRulesetContainer(rulesetContainer);
+        }
+
+        private class TaikoFlashlight : Flashlight
+        {
+            private readonly TaikoPlayfield taikoPlayfield;
+
+            public TaikoFlashlight(TaikoPlayfield taikoPlayfield)
+            {
+                this.taikoPlayfield = taikoPlayfield;
+                FlashlightSize = new Vector2(0, getSizeFor(0));
+            }
+
+            private float getSizeFor(int combo)
+            {
+                if (combo > 200)
+                    return default_flashlight_size * 0.8f;
+                else if (combo > 100)
+                    return default_flashlight_size * 0.9f;
+                else
+                    return default_flashlight_size;
+            }
+
+            protected override void OnComboChange(int newCombo)
+            {
+                this.TransformTo(nameof(FlashlightSize), new Vector2(0, getSizeFor(newCombo)), FLASHLIGHT_FADE_DURATION);
+            }
+
+            protected override string FragmentShader => "CircularFlashlight";
+
+            public override bool Invalidate(Invalidation invalidation = Invalidation.All, Drawable source = null, bool shallPropagate = true)
+            {
+                if ((invalidation & Invalidation.DrawSize) > 0)
+                {
+                    Schedule(() =>
```

Comment:
This is kind-of a new pattern that we haven't used elsewhere, and may lead to issues such as multiple invalidations per-frame causing multiple schedules.

Better than this is to use a `Cached` value which you validate inside `Update()` - check how it's done in `SpriteText` / `FlowContainer`, etc for reference.